`repo[ref]` raises an `AssertionError` when ref has length 32

https://github.com/jelmer/dulwich/blob/09d6131c31ed8c7046ff185bb2a57ad2bf8cdbba/dulwich/repo.py#L1200 

```python
from dulwich.repo import Repo
from dulwich.porcelain import branch_create

repo = Repo(".")
branch_create(repo, b"feat-backend-refactor")
repo[b"refs/heads/feat-backend-refactor"]
```

```pytb
----> 6 repo[b"refs/heads/feat-backend-refactor"]

File ~/projects/dvcorg/scmrepo/.venv/lib/python3.14/site-packages/dulwich/repo.py:1204, in BaseRepo.__getitem__(self, name)
   1201 if len(name) in (20, 32, 40, 64):  # Support both SHA1 and SHA256
   1202     try:
   1203         # Try as ObjectID/RawObjectID
-> 1204         return self.object_store[
   1205             ObjectID(name) if len(name) == 40 else RawObjectID(name)
   1206         ]
   1207     except (KeyError, ValueError):
   1208         pass

File ~/projects/dvcorg/scmrepo/.venv/lib/python3.14/site-packages/dulwich/object_store.py:384, in BaseObjectStore.__getitem__(self, sha1)
    382 def __getitem__(self, sha1: ObjectID | RawObjectID) -> ShaFile:
    383     """Obtain an object by SHA1."""
--> 384     type_num, uncomp = self.get_raw(sha1)
    385     return ShaFile.from_raw_string(
    386         type_num, uncomp, sha=sha1, object_format=self.object_format
    387     )

File ~/projects/dvcorg/scmrepo/.venv/lib/python3.14/site-packages/dulwich/object_store.py:2266, in DiskObjectStore.get_raw(self, name)
   2263             pass
   2265 # Fall back to the standard implementation
-> 2266 return super().get_raw(name)

File ~/projects/dvcorg/scmrepo/.venv/lib/python3.14/site-packages/dulwich/object_store.py:1196, in PackBasedObjectStore.get_raw(self, name)
   1194     hexsha = None
   1195 else:
-> 1196     raise AssertionError(f"Invalid object name {name!r}")
   1197 for pack in self._iter_cached_packs():
   1198     try:

AssertionError: Invalid object name b'refs/heads/feat-backend-refactor'
```


Looks like a regression from https://github.com/jelmer/dulwich/pull/1604.

--- a/NEWS
+++ b/NEWS
@@ -1,5 +1,8 @@
 0.25.1	UNRELEASED
 
+ * Fix AssertionError when accessing ref names with length matching binary
+   hash length (e.g., 32 bytes for SHA-256). (Jelmer Vernooĳ, #2040)
+
  * Add ``parse_commit_broken`` function to parse broken commits.
    (Valentin Lorentz, Jelmer Vernooĳ)
 

--- a/dulwich/repo.py
+++ b/dulwich/repo.py
@@ -1197,18 +1197,27 @@ class BaseRepo:
         """
         if not isinstance(name, bytes):
             raise TypeError(f"'name' must be bytestring, not {type(name).__name__:.80}")
-        if len(name) in (20, 32, 40, 64):  # Support both SHA1 and SHA256
+        # If it looks like a ref name, only try refs
+        if name == b"HEAD" or name.startswith(b"refs/"):
             try:
-                # Try as ObjectID/RawObjectID
+                return self.object_store[self.refs[Ref(name)]]
+            except (RefFormatError, KeyError):
+                pass
+        # Otherwise, try as object ID if length matches
+        if len(name) in (
+            self.object_store.object_format.oid_length,
+            self.object_store.object_format.hex_length,
+        ):
+            try:
                 return self.object_store[
-                    ObjectID(name) if len(name) == 40 else RawObjectID(name)
+                    ObjectID(name)
+                    if len(name) == self.object_store.object_format.hex_length
+                    else RawObjectID(name)
                 ]
             except (KeyError, ValueError):
                 pass
-        try:
-            return self.object_store[self.refs[Ref(name)]]
-        except RefFormatError as exc:
-            raise KeyError(name) from exc
+        # If nothing worked, raise KeyError
+        raise KeyError(name)
 
     def __contains__(self, name: bytes) -> bool:
         """Check if a specific Git object or ref is present.

--- a/tests/test_repository.py
+++ b/tests/test_repository.py
@@ -292,6 +292,17 @@ class RepositoryRootTests(TestCase):
 
         self.assertRaises(ValueError, r.__delitem__, b"notrefs/foo")
 
+    def test_getitem_32_byte_ref(self) -> None:
+        """Test that accessing a ref name that's 32 bytes long works (issue #2040)."""
+        r = self.open_repo("a.git")
+        # Create a ref with exactly 32 bytes
+        ref_name = b"refs/heads/feat-backend-refactor"
+        self.assertEqual(len(ref_name), 32)
+        r[ref_name] = b"a90fa2d900a17e99b433217e988c4eb4a2e9a097"
+        # This should not raise AssertionError
+        obj = r[ref_name]
+        self.assertEqual(obj.id, b"a90fa2d900a17e99b433217e988c4eb4a2e9a097")
+
     def test_get_refs(self) -> None:
         r = self.open_repo("a.git")
         self.assertEqual(

